magma/smart/smart_bits.py

```diff
@@ -537,4 +556,47 @@ def _eval(lhs: SmartBits, rhs: _SmartExpr) -> (SmartBits, _SmartExpr):
     rhs = copy.deepcopy(rhs)
     rhs.resolve(Context(lhs, rhs))
     res = rhs.eval()
-    return SmartBits.make(res), rhs
+    return SmartBits.from_bits(res), rhs
+
+
+def eval(expr: _SmartExpr, width: int, signed: bool = False):
+    lhs = SmartBits[width, signed]()
+    lhs @= expr
+    return lhs
+
+
+class _SmartifyTypeTransformer(TypeTransformer):
+    def visit_Bits(self, T):
+        signed = issint(T)
+        return SmartBits[len(T), signed]
+
+
+class _LeafCollector(ValueVisitor):
+    def __init__(self):
+        self.bits_leaves = []
+        self.other_leaves = []
+
+    def visit_Digital(self, value):
+        self.other_leaves.append(value)
+
+    def visit_Bits(self, value):
+        self.bits_leaves.append(value)
+
+
+def make_smart(value):
+    T = type(value)
+    Tsmart = _SmartifyTypeTransformer().visit(T)
+    Tsmart = Tsmart.qualify(Direction.Undirected)
+    smart_value = Tsmart()
+    leaf_collector = _LeafCollector()
+    leaf_collector.visit(value)
+    for bits in leaf_collector.bits_leaves:
+        selector = make_selector(bits)
+        smart_bits = selector.select(smart_value)
+        smart_bits_unwrapped = smart_bits._get_magma_value_()
+        smart_bits_unwrapped @= bits
+    for leaf in leaf_collector.other_leaves:
+        selector = make_selector(leaf)
+        new_leaf = selector.select(smart_value)
+        new_leaf @= leaf
```

Comment:
Do these need to be separate loops? If wiring handles the fetching of the magma protocol value, perhaps you don't need to handle smart bits separately?

Reply:
Good point. I needed to augment `SmartBits.wire` to handle this case, which it should have.